pkg/daemon/server/daemon_server.go

```diff
@@ -87,6 +89,14 @@ func (ds *daemonServer) Run(ctx context.Context) error {
 		}
 	}()
 
+	rateCalculators := make(map[string]*server.RateCalculator, len(ds.pipeline.Spec.Vertices))
```

Comment:
Can we use one single `rateCalculator` for all the vertices? And we just new a calculator here, and pass it to `newGRPCServer()`.

The place to start the calculator should be before `<-ctx.Done()`.